gstatus.8
.\" Man page for gstatus
.\" Contact pcuzner@redhat.com to correct errors or typos
.TH gstatus 8 "08 July 2015" "0.64-2" "gstatus man page"
.SH NAME
gstatus \- query the health of a glusterfs trusted storage pool
.SH SYNOPSIS
\fBgstatus\fR [-a|--all] [-b|--backlog] [-D|--debug] [-l|--layout] [-o|--output-mode json|keyvalue|\fIconsole\fR] [-s|--state] [-t|--timeout secs] [-u|--units dec|\fIbin\fR] [-v|--volume \fIALL\fR|volume name] [-w|--without-progress]
.SH DESCRIPTION
CLI command to provide a status check on the cluster’s health, providing
a view of node, volume state (up, degraded, partial or down), brick 
up/down states and capacity information by volume (usable, used).
.PP
In addition to the interactive model, a json or keyvalue output option 
is also available through '-o json|keyvalue'. By utilising -o, you can 
log the state of the cluster to a file, and interpret with Splunk, 
Logstash or nagios/zabbix.
.PP 
Errors detected, are listed in simple terms and provide an easy way to 
assess the impact to a service, following a disruptive event that
affects the trusted pool.

.SH OPTIONS
.PP
.B -a, --all
.RS
provides the same output as an invocation using -s and -v
.RE
.PP
.B -b, --backlog
.RS
by default the list of files waiting to be healed is not checked since the command to determine this state may take several seconds to complete. If an accurate view of self heal state needs to be included in the output, use -b to trigger the additional command "vol heal <vol> info" to be executed.
.RE
.PP
.B -D, --debug
.RS
enable more verbose output to aid debugging
.RE
.PP
.B -l, --layout
.RS
shows the physical layout of the volume by brick, making it easier to understand the relationships between servers and bricks
.RE
.PP
.B -o, --output-mode \fRjson | keyvalue | \fIconsole\fR
.RS
control how the data gathered by the tool is presented back to the user. The json and keyvalue options allow a summary of the clusters state to be captured in a format that could be recorded in a log file for trending, or analysis purposes. When this option is not specfied, the default output mode is 'console'
.RE
.PP
.B -s, --state
.RS
show the high level state of the trusted storage pool, summarising node counts and capacity utilisation data
.RE
.PP
.B -t, --timeout\fR secs
.RS
gstatus issues multiple commands to gluster to gather state data. On busy systems these commands could take several seconds, which can make the tool seem unresponsive. Rather than force the admin to wait for prolonged periods, a command timeout is enforced. By default gstatus uses a timeout of \fB5\fR seconds (set in gstatuscfg/config), but this may be increased using the '-t' option
.RE
.PP
.B -u, --units  \fRdec|\fIbin\fR
.RS
capacity can be calculated either in decimal or binary representations (TB vs TiB). The default is to provide all capacity information in binary format (i.e. GiB, TiB), which matches the methods used by tools such as 'df'.
.RE
.PP
.B -v, --volume \fIALL\fR | volume-name
.RS
Provide detailed information about the volume including
.RE
.PP
.B -w, --without-progress
.RS
turn of the progress text normally shown as data is gathered and assembled
.RE
.SH EXAMPLES
.B gstatus -a
.br
Gather and present cluster and volume information
.PP 
.B gstatus -al
.br
Show cluster and volume information, and for each volume show the brick layout which includes individual capacity utilisation information
.PP
.B gstatus -o keyvalue
.br
Gather the data and format the results into a keyvalue string. Useful to capture point-in-time configuration information for capacity trending or forwarding to other systems for health monitoring.
.SH SEE ALSO
gluster(8)
.SH BUGS
1) gstatus supports distributed, replicated and disperse volume types only
.PP
2) gluster vol status does not report on self heal heal for disperse volumes, so gstatus is unable to provide an accurate summary of self heal state for this volume type
.PP
3) Normally every node will be capacity nodes. However, when nodes are added to the cluster for quorum purposes, the Self Heal field shows more nodes with Self Heal active than needed. This is due to the arbitration nodes also running the self head daemon.
.SH AUTHOR
Paul Cuzner (pcuzner@redhat.com)


gstatus.py
#!/usr/bin/env python


from optparse import OptionParser  # command line option parsing
from datetime import datetime

import gstatus.gstatuscfg.config as cfg
from gstatus.libutils.utils import display_bytes, version_ok
from gstatus.libgluster.cluster import Cluster


def console_mode():
    """ Produce the output to the users console - stdout """

    if cluster.messages:
        status_msg = "%s(%d)" % (cluster.status.upper(), len(cluster.messages))
    else:
        status_msg = cluster.status.upper()

    # General Status Header
    print ("     Product: %s  Capacity:%s(raw bricks)" % (cluster.product_shortname.ljust(17),
                                                          display_bytes(cluster.raw_capacity, display_units).rjust(11)))

    print ("      Status: %s           %s(raw used)" % (status_msg.ljust(17),
                                                        display_bytes(cluster.used_capacity, display_units).rjust(11)))

    print ("   Glusterfs: %s%s%s(usable from volumes)" % (cluster.glfs_version.ljust(17),
                                                          " " * 11,
                                                          display_bytes(cluster.usable_capacity, display_units).rjust(
                                                              11)))

    if cluster.snapshot_capable:
        snap_msg = "Snapshots: %s" % (str(cluster.snapshot_count).rjust(3))
    else:
        snap_msg = ""

    print ("  OverCommit: %s%s%s" % (cluster.over_commit.ljust(3), " " * 15, snap_msg))

    # Component Status
    if state_request:
        print ("\n   Nodes       :%3d/%3d\t\t  Volumes:  %2d Up"
               % (cluster.nodes_active, cluster.node_count,
                  cluster.volume_summary['up']))

        print ("   Self Heal   :%3d/%3d\t\t            %2d Up(Degraded)"
               % (cluster.sh_active, cluster.sh_enabled,
                  cluster.volume_summary['degraded']))

        print ("   Bricks      :%3d/%3d\t\t            %2d Up(Partial)"
               % (cluster.bricks_active, cluster.brick_count,
                  cluster.volume_summary['partial']))

        print ("   Connections :%3d/%4d%s%2d Down" %
               (cluster.num_clients,
                cluster.num_connections,
                " " * 20,
                cluster.volume_summary['down']))

    # Volume Breakdown
    if volume_request:
        print "\nVolume Information"

        for vol_name in sorted(cluster.volume):

            if len(volume_list) == 0 or vol_name in volume_list:
                vol = cluster.volume[vol_name]
                (up_bricks, all_bricks) = vol.brick_states()

                task_list_str = ", ".join(vol.task_list) if vol.task_list else 'None'

                print ("\t%s %s - %d/%d bricks up - %s"
                       % (vol_name.ljust(16, ' '),
                          vol.volume_state.upper(),
                          up_bricks, all_bricks,
                          vol.typeStr))

                print ("\t" + " " * 17 +
                       "Capacity: (%d%% used) %s/%s (used/total)"
                       % (vol.pct_used,
                          display_bytes(vol.used_capacity, display_units),
                          display_bytes(vol.usable_capacity, display_units)))

                if cluster.snapshot_capable:
                    print "\t" + " " * 17 + "Snapshots: %d" % vol.snapshot_count

                print ("\t" + " " * 17 + "Self Heal: %s" % vol.self_heal_string)
                print ("\t" + " " * 17 + "Tasks Active: %s" % task_list_str)
                print ("\t" + " " * 17 + "Protocols: glusterfs:%s  NFS:%s  SMB:%s"
                       % (vol.protocol['NATIVE'],
                          vol.protocol['NFS'],
                          vol.protocol['SMB']))
                print ("\t" + " " * 17 +
                       "Gluster Connectivty: %d hosts, %d tcp connections"
                       % (vol.num_clients,
                          vol.num_connections))

                if volume_layout:
                    print
                    vol.print_layout()

                print

    if state_request:
        print "\nStatus Messages"
        if cluster.messages:

            # Add the current cluster state as the first message to display
            cluster.messages.insert(0, "Cluster is %s" % cluster.status.upper())
            for info in cluster.messages:
                print "  - " + info

        else:
            print "  - Cluster is HEALTHY, all_bricks checks successful"

    print


def log_mode():
    """ produce the output suitable for later processing by logstash, 
        or splunk et al """

    now = datetime.now()
    print "%s %s" % (now, str(cluster))


def main():
    if cluster.output_mode == 'console':
        # add some spacing to make the output stand out more
        print " "

        # setup up the cluster object structure
    cluster.initialise()

    # run additional commands to get current state
    cluster.update_state(self_heal_backlog)

    # use the bricks to determine overall cluster disk capacity
    cluster.calc_capacity()

    # perform checks on the clusters state
    cluster.health_checks()

    # Now with the object model complete, we can provide the info
    # to the user
    if cluster.output_mode == 'console':
        console_mode()

    elif cluster.output_mode in ['json', 'keyvalue']:
        log_mode()


if __name__ == '__main__':

    # initialise any global variables
    cfg.init()

    usageInfo = "usage: %prog [options]"

    parser = OptionParser(usage=usageInfo, version="%prog 0.64")
    parser.add_option("-s", "--state", dest="state", action="store_true", help="show highlevel health of the cluster")
    parser.add_option("-v", "--volume", dest="volumes", action="store_true",
                      help="volume info (default is ALL, or supply a volume name)")
    parser.add_option("-b", "--backlog", dest="selfheal", action="store_true", default=False,
                      help="Look deeper at self heal state")
    parser.add_option("-a", "--all", dest="everything", action="store_true", default=False,
                      help="show all cluster information (-s with -v)")
    parser.add_option("-u", "--units", dest="units", choices=['bin', 'dec'],
                      help="display capacity units in DECimal or BINary format (GB vs GiB)")
    parser.add_option("-l", "--layout", dest="layout", action="store_true", default=False,
                      help="show brick layout when used with -v, or -a")
    parser.add_option("-o", "--output-mode", dest="output_mode", default='console',
                      choices=['console', 'json', 'keyvalue'],
                      help="produce output in different formats - json, keyvalue or console(default)")
    parser.add_option("-D", "--debug", dest="debug_on", action="store_true", help="turn on debug mode")
    parser.add_option("-w", "--without-progress", dest="progress", action="store_true", default=False,
                      help="turn off progress updates to user during data gathering")
    parser.add_option("-t", "--timeout", dest="timeout", help="gluster command timeout value (secs)")
    (options, args) = parser.parse_args()

    if options.timeout:
        cfg.CMD_TIMEOUT = int(options.timeout)

    cfg.debug = True if options.debug_on else False

    cfg.no_progress_msgs = options.progress

    state_request = options.state

    volume_request = options.volumes

    volume_layout = options.layout

    self_heal_backlog = options.selfheal

    display_units = options.units if options.units else 'bin'

    volume_list = []  # empty list of vols = show them all

    # default behaviours
    if volume_request and args:
        volume_list = args

    if state_request:
        self_heal_backlog = False

    if options.everything:
        state_request = True
        volume_request = True

    if options.output_mode != 'console':
        self_heal_backlog = False

    # Create a cluster object. This simply creates the structure of 
    # the object and populates the glusterfs version 
    cluster = Cluster()

    cluster.output_mode = options.output_mode

    if version_ok(cluster.glfs_version, cfg.min_version):

        main()

    else:

        print "gstatus is not compatible with this version of glusterfs %s" % cluster.glfs_version
        exit(16)
